### ISSUE TITLE
Permissions not saved for one roles (i have 5 roles)

### ISSUE THREAD

**Issue by SunD004** [strapi/strapi#8511]
### Describe the bug

When i save permissions for a role and i refresh my page permissions is not saved

### Steps to reproduce the behavior

1. Go to Settings
2. Click on Roles
3. Choose a Role
4. Moidify a permisisons & save
5. Refresh and see permissions not saved

### Expected behavior

I want permissions saving.
I have 4 roles and only 1 role is affected by this bug

### Screenshots

![Screenshot from 2020-10-29 16-40-19](https://user-images.githubusercontent.com/44288530/97596901-84cb3280-1a05-11eb-8f9a-668349cd448c.png)

![Screenshot from 2020-10-29 16-40-26](https://user-images.githubusercontent.com/44288530/97596932-8eed3100-1a05-11eb-8dd6-c44c8e0beb38.png)

### System

- Node.js version: v12.16.0
- NPM version: 6.13.4
- Strapi version: V3.2.5
- Database: SQLITE
- Operating system: Ubuntu LINUX

**Comment by derrickmehaffy** [comment]
`Database: GraphQL` GraphQL in this case isn't a database, can you confirm which database you are using?

**Comment by SunD004** [comment]
> `Database: GraphQL` GraphQL in this case isn't a database, can you confirm which database you are using?

OVH Database / GraphQL

**Comment by derrickmehaffy** [comment]
`OVH Database / GraphQL`

Still not sure I understand is this:
- MongoDB
- MySQL / MariaDB
- PostgreSQL
- SQLite

These are the only databases we support, GraphQL is a request layer similar to REST.

**Comment by SunD004** [comment]
> `OVH Database / GraphQL`
> 
> Still not sure I understand is this:
> 
> * MongoDB
> * MySQL / MariaDB
> * PostgreSQL
> * SQLite
> 
> These are the only databases we support, GraphQL is a request layer similar to REST.

I think it's MongoDB.
But it's not a database problem it's a strapi problem... All my roles work great except one..

**Comment by derrickmehaffy** [comment]
In this instance @SunD004 we need steps to reproduce the issue to determine if it's a code issue or a database problem.

If possible can you make an export of the users-permissions collections in your database (omiting the user one as to not dump any private information) and do you happen to have a public git repo for your project so I can try to reproduce locally?

**Comment by SunD004** [comment]
> In this instance @SunD004 we need steps to reproduce the issue to determine if it's a code issue or a database problem.
> 
> If possible can you make an export of the users-permissions collections in your database (omiting the user one as to not dump any private information) and do you happen to have a public git repo for your project so I can try to reproduce locally?

Hello,
I created a repo with the project who have the problem of saving permissions
Link => https://github.com/SunD004/strapi
Thank for your help

**Comment by derrickmehaffy** [comment]
Will do some testing thanks @SunD004

**Comment by derrickmehaffy** [comment]
@SunD004 Alright I think I have identified the problem, however what is not clear is how it happened as I cannot reproduce it other than with your example. I have worked with our lead backend engineer and we think the cause is from your bootstrap.js file where you are pre-populating some entries.

The specific error is related to duplicate permission keys in your database:
![image](https://user-images.githubusercontent.com/8593673/97729720-3616aa80-1a90-11eb-95cf-1abbbfbe9057.png)

We have code to ensure this doesn't happen when Strapi itself creates the permissions **on boot of the application** but not any validation after they have been initialized.

The specific problem comes from here: https://github.com/strapi/strapi/blob/717a834a045bb2cbae54526fcd429546f5b53642/packages/strapi-plugin-users-permissions/services/UsersPermissions.js#L199-L204

It's fetching all the permissions and reducing them, during this reduce the actual permissions that were updated are being removed from the response.

---

In short we do need to add some code to remove duplicate permissions but for your very specific issue, we think the root cause is coming from your bootstrap.js file. If you could explain what you are doing here and why just to see if we can help you construct a better way to prevent it until we can add the duplication cleanup code.

In the short term for you to fix this and cleanly remove the duplicate permissions, we think that deleteing and recreating the role should work and remove the duplicates. Else you may need to construct a .sql script to clean them up.

**Comment by derrickmehaffy** [comment]
Adding a comparison of what the permissions should look like for a role/model (this is also from your same example application)

![image](https://user-images.githubusercontent.com/8593673/97730284-e71d4500-1a90-11eb-8c6c-0be56e4c0bcf.png)

**Comment by aksdevac** [comment]
I have also faced this 'duplicate permissions' issue when I add new controller or rename controller in some prev. strapi version on prod env with pm2, but not tried to debug it. I just cleaned up permissions, but I think it was due to cluster mode of pm2, if I take pull on prod server & restart pm2 then is there any chance that all instance will create new entry for renamed or added controller ?

**Comment by derrickmehaffy** [comment]
That is a great example @aksdevac it is possible that the cluster throws the permission creation into a race condition. It would be very difficult to replicate reliably but is certainly possible.

@alexandrebodin I think this likely enforces that we need duplicate permission checking and cleanup, or possibly just the option to skip all the bootup checking options (or both)

**Comment by jalal246** [comment]
This may be related to this issue #8528.

**Comment by kevinvugts** [comment]
This one happend to me when upgrading to 3.4.4.

What is this fix for this @derrickmehaffy ?

**Comment by derrickmehaffy** [comment]
> This one happend to me when upgrading to 3.4.4.
> 
> What is this fix for this @derrickmehaffy ?
> 

Currently you have to cleanup the database manually

**Comment by manutepowa** [comment]
> I have also faced this 'duplicate permissions' issue when I add new controller or rename controller in some prev. strapi version on prod env with pm2, but not tried to debug it. I just cleaned up permissions, but I think it was due to cluster mode of pm2, if I take pull on prod server & restart pm2 then is there any chance that all instance will create new entry for renamed or added controller ?

In my case, have a duplicate permissions. 
I also use PM2 configured as a cluster with 2 threads.
### System
Node.js version: v14.17.6
NPM version: 6.14.15
Strapi version: 3.6.8
Database: MYSQL
PM2: 4.5.0

**Comment by jakubmirejovsky** [comment]
Any workaround? I am thinking about NOT running pm2 restart all after git pull & npm run build, instead run npm run dev? Si the DB update process happens only once?

I can confirm that there are as many duplicated entries as is the pm2 cluster instances count.

**Comment by sircovsw** [comment]
Any ideas?

**Comment by derrickmehaffy** [comment]
can't reproduce in v4, marking as closed
